core/types.ml

```diff
@@ -2401,6 +2470,964 @@ struct
+      = if Settings.get trace_printer
+        then
+          begin
+            fun bt s ->
+            bt.trace <- (s, bt.current_printer) :: bt.trace;
+          end
+        else (fun {buf;_} s -> Buffer.add_string buf s)
+
+    let apply : 'a printer -> Context.t -> 'a -> t -> unit
+      = if Settings.get trace_printer
+        then
+          (fun pr ctx thing bt ->
+            match pr with
+            | Empty -> ()
+            | Printer (n, pr) ->
+               let upper = bt.current_printer in
+               bt.current_printer <- n;
+               pr ctx thing bt;
+               bt.current_printer <- upper)
+        else
+          (fun pr ctx thing buf ->
+            match pr with
+            | Empty -> ()
+            | Printer (_, pr) -> pr ctx thing buf)
+
+    let concat : sep:string -> 'a printer -> 'a list -> Context.t -> t -> unit
+      = fun ~sep pr items ctx buf ->
+      match pr with
+      | Empty -> ()
+      | Printer _ ->
+         let rec loop
+           = fun sep pr items ctx buf ->
+           begin
+             match items with
+             | [] -> ()
+             | [last] -> apply pr ctx last buf
+             | not_last :: rest ->
+                apply pr ctx not_last buf;
+                write buf sep;
+                loop sep pr rest ctx buf
+           end
+         in
+         loop sep pr items ctx buf
+
+    let concat' : sep:string -> unit printer list -> Context.t -> t -> unit
+      = fun ~sep prs ctx buf ->
+      let rec loop sep ctx buf = function
+        | [] -> ()
+        | [pr] -> apply pr ctx () buf
+        | pr :: prs ->
+           apply pr ctx () buf;
+           write buf sep;
+           loop sep ctx buf prs
+      in
+      loop sep ctx buf (List.filter (function Empty -> false | _ -> true) prs)
+
+    let seq : sep:string -> ('a printer * 'b printer) -> ('a * 'b) -> Context.t -> t -> unit
+      = fun ~sep (lp, rp) (x, y) ctx buf ->
+      match lp, rp with
+      | Printer _, Printer _ ->
+         apply lp ctx x buf;
+         write buf sep;
+         apply rp ctx y buf
+      | Printer _, Empty ->
+         apply lp ctx x buf
+      | Empty, Printer _ ->
+         apply rp ctx y buf
+      | Empty, Empty -> ()
+
+    let constant : string -> unit printer
+      = fun s ->
+      Printer ("<constant>", fun _ctx () buf -> write buf s)
+
+    let wrap : 'a printer -> 'a -> unit printer
+      = fun pr v ->
+      Printer ("<wrap>", fun ctx () buf -> apply pr ctx v buf)
+
+    let wrap_ambient : Context.ambient -> 'a printer -> 'a -> unit printer
+      = fun amb pr v ->
+      Printer ("<wrap>", fun ctx () buf ->
+                         let inner_ctx = Context.set_ambient amb ctx in
+                         apply pr inner_ctx v buf)
+
+    let eval : 'a printer -> Context.t -> 'a -> string
+      = if Settings.get trace_printer
+        then
+          (fun pr ctx v ->
+            let buf = create 10 in (* initial size is an arbitrary value here (TODO) *)
```

Comment:
Make the initial size much bigger, let's arbitrarily say, the maximum length of a tweet: 280. The initial allocation is cheap, subsequent reallocations are more expensive, so you generally want to avoid too many reallocations.